async/actorpromise.md

```diff
@@ -0,0 +1,127 @@
+
+Getting better, but not good enough. What we _really_ want to be able to do is query multiple actors to get the account summary data and then send _all_ of that data (preferably bundled up in a nice array) to a destination actor that can then display and/or process the information. For this we're going to need an intermediary - something that awaits promise fulfillment and adds to a collection when fulfilled. Finally, once this intermediary has received all expected fulfillments, it can then send the results to a destination. You can think of this intermediary as a _promise buffer_:
+
+```pony
+interface CollectionReceiver[A]
+  be receivecollection(coll: Array[A] val)
+  
+actor Collector[A: Any #send]  
+  var _coll: Array[A] iso
+  let _max: USize
+  let _target: CollectionReceiver[A] tag
+  
+  new create(max: USize, target: CollectionReceiver[A] tag) =>    
+    _coll = recover Array[A] end
+    _max = max
+    _target = target
+    
+  be receive(s: A) =>
+    _coll.push(consume s)    
+    if _coll.size() == _max then      
+      let output = _coll = recover Array[A] end
+      _target.receivecollection(consume output)
+    end
+```
+Now that we have an intermediary, we can create multiple promises to send to multiple bank accounts:
+```pony
+let accounts = ["0001"; "0002"; "0003"; "0004"]
+let collector = Collector[AccountSummary](accounts.size(), this)
+    
+for account in accounts.values() do
+  let aggregate = AccountAggregate(account, 6000)
+  let p = Promise[AccountSummary]
+  p.next[None](recover collector~receive() end)
+  aggregate.summarize(p)
+end 
+```
+Our bank account aggregate can be modified to include an account summary with the **summarize** method:
+
+```pony
+be summarize(p: Promise[AccountSummary]) =>    
+  p(recover AccountSummary(_balance, _account) end)
+```
+
+Finally, we add the behavior that the collector is expecting to the calling actor in order to respond to the list of account summaries:
+
+```pony
+be receivecollection(coll: Array[AccountSummary] val) =>
+  _env.out.print("received account summaries:")
+  for summary in coll.values() do
+    _env.out.print("Account " + summary.accountnumber() + ": $" + summary.currentbalance().string())
+  end
+```
+## Discussion
+Actor systems have been around for quite some time now, but most developers don't default to modeling their problems as actor patterns. Most of us want to solve this problem with synchronous code that looks like this:
+
+```pony
+for acct in _accounts.values() do
+    _summaries.push(acct.summarize())
+end
+```
+Truthfully this is probably a good way to solve a lot of problems. In the trivial example discussed here, we're not actually gaining all that much from the use of actors, so why incur all that extra complexity? Consider these questions:
```

Comment:
I think this discussion suggests that there could be a better example to use to show the value of using promise to collect values from a number of actors. This feels a little forced. I think the big value in the pattern could be obscured below the "why would I do it that way?"